# Results and Discussion

## Scalability, Portability, and Efficiency

The reimplementation of EAGER into Nextflow offers a range of benefits over the original custom pipeline framework.

Firstly, the new framework provides immediate integration of nf-core/eager into various schedulers in POSIX High-Performance-Cluster (HPC) environments, cloud computing resources, and as well as local workstations. This portability allows both small and big labs to run nf-core/eager regardless of the type of computer or cluster size, with minimal effort or configuration, facilitating reproducibility and therefore maintenance of standards within the field. This is further assisted by the in-built compatibility with software environments and containers such as conda, docker and singularity. This ensures exact versions of software are used by a user, regardless of the setup of their local software environment. Another major change with nf-core/eager is that the GUI input is now replaced with a command-line-interface as the primary user interaction mode. This is more compatible and portable with most HPCs (that may not offer X11 forwarding), and is in line with the vast majority of bioinformatic tools. We therefore believe this will not be a hindrance to new researchers from outside computational biology. However there are plans within the nf-core community to provide multiple alternatives in the near future including a command-line wizard and a web-based input GUI.

Secondly, reproducibility is made easier through the use of 'profiles' that can define configuration parameters. These profiles can be managed at different hierarchical levels. HPC-level profiles can specify parameters for the computing environment (schedulers, cache locations, maximum resource etc.), which can be centrally managed to ensure all users of a group use the same settings. Pipeline-level profiles, specifying default parameters for nf-core/eager itself, allow fast access to setups via a single flag for routine analyses, without having to newly configure each run. Compared to the original EAGER that utilised per-FASTQ XML files with hardcoded filepaths, nf-core/eager allows researchers to publish the specific profile used in their runs alongside their publications, to ensure other groups can generate the same results. Usage of profiles also reduces mistakes caused by insufficient 'prose' based reporting of program settings, which may occur in papers written by researchers unfamiliar with informatics. The default nf-core/eager profile uses parameters evaluated in different aDNA specific contexts (e.g. @doi:10.1186/1471-2164-13-178), and will be updated in each new release as new studies are published.

nf-core/eager provides improved efficiency over the original EAGER pipeline by replacing the sample-by-sample sequential processing with Nextflow's asynchronous parallelisation. This, combined with pre-defined per-process customisation of resource parameters, reduces unnecessary resource allocation that can occur with new users to each step of an NGS data processing pipeline. This is particularly pertinent given the increasing use of centralised HPCs or cloud computing, which often use per-hour cost calculations.

## Latest aDNA practices

nf-core/eager follows a similar structural foundation with the original EAGER. Given Illumina short-read FASTQ and/or BAM files, and a reference FASTA file, this can be split into four main stages:

1. Preprocessing
   - Sequencing quality control: FastQC (@url:https://www.bioinformatics.babraham.ac.uk/projects/fastqc/)
   - Sequencing artefact clean-up: AdapterRemoval2 (@doi:10.1186/s13104-016-1900-2)
   - Preprocessing statistics generation
2. Mapping and post- processing
   - Alignment against reference genome: BWA (@doi:10.1093/bioinformatics/btp324, @doi:10.1093/bioinformatics/btp698, @arXiv:1303.3997), CircularMapper (@doi:10.1186/s13059-016-0918-z)
   - Mapping quality filtering: Samtools (@doi:10.1093/bioinformatics/btp352)
   - PCR duplicate removal: DeDup (@doi:10.1186/s13059-016-0918-z), Picard MarkDuplicates(@doi:10.1101/gr.107524.110)
   - Mapping statistics generation: PreSeq (@doi:10.1038/nmeth.2375), Qualimap2(@doi:10.1093/bioinformatics/btv566)
3. aDNA Evaluation and Modification
   - Damage profiling: DamageProfiler (@doi:10.5281/zenodo.3557708)
   - aDNA reads selection: PMDTools (@doi:10.1073/pnas.1318934111)
   - Damage removal: Bamutils(@doi:10.1101/gr.176552.114)
   - (Human) contamination estimation: ANGSD (@doi:10.1186/s12859-014-0356-4)
   - Mitochondrial to Nuclear Ratio: MTNucRatio (@doi:10.1186/s13059-016-0918-z)
4. Genotyping and Consensus Sequencing: GATK (@doi:10.1101/gr.107524.110), VCF2Genome (@doi:10.1186/s13059-016-0918-z)

In nf-core/eager, all tools also originally used in EAGER have been updated to latest versions, as available on Bioconda (@doi:10.1038/s41592-018-0046-7) and conda-forge (@url:https://conda-forge.github.io) to ensure widespread accessibility and stability of utilized tools. The MapDamage2 (@doi:10.1093/bioinformatics/btt193) and Schmutzi (@doi:10.1186/s13059-015-0776-0) methods have not been carried over to nf-core/eager, the first because a more performant successor method (DamageProfiler) exists, and the latter because a stable release of the method could not be migrated to Bioconda. We anticipate that there will be an updated version of Schmutzi in the near future that will allow us to integrate the method again in nf-core/eager, once a version is released on Bioconda. Support for Bowtie2 (@doi:10.1038/nmeth.1923) will be added in the near future, after consultation with the palaeogenetics community. New tools to the basic workflow include fastp (@doi:10.1093/bioinformatics/bty560) for the removal of poly-G sequencing artefacts that are common in 2-colour Illumina sequencing machines (such as the increasingly popular NextSeq and NovaSeq platforms). We have also included the FreeBayes genotyper as an alternative to the human-focused GATK tools. We have also maintained the possibility of using the now unsupported GATK UnifiedGenotyper, as the GATK HaplotypeCaller performs _de novo_ assembly around possible variants, which may not be suitable for low-coverage aDNA data.

We have further extended functionality of the pipeline, primarily focusing on ancient metagenomic analysis to be run alongside the standard genomic analysis against a single reference genome. We have added the ability to screen all off-target reads (from mapping to the supplied reference FASTA) with two metagenomic profilers: MALT (@doi:10.1101/050559, @doi:10.1038/s41559-017-0446-6) and Kraken2 (@doi:10.1186/s13059-019-1891-0). Characterisation of properties of authentic ancient DNA (@doi:10.1073/pnas.0704665104) from MALT alignments is carried out with the HOPS pipeline (@doi:10.1186/s13059-019-1903-0). Ancient metagenomic studies sometimes may include comparative samples from living individuals (@doi:10.1186/s40168-019-0717-3). To support open data, whilst respecting data privacy, nf-core/eager includes a 'strip_fastq' script which creates raw FASTQ files but with reference-genome mapped reads removed. This then allows safe upload of sequencing data to public repositories with identifiable human data removed.

Additional functionality tailored for ancient bacterial genomics includes integration of a SNP alignment generation tool, MultiVCFAnalyzer (@doi:10.1038/nature13591), which allows assessment of levels of cross-mapping from different related taxa to a reference genome  - a common challenge in ancient bacteria genome reconstruction (@doi:10.1146/annurev-genom-091416-035526). Simple coverage statistics of particular annotations (e.g. genes) of an input reference is offered by bedtools (@doi:10.1093/bioinformatics/btq033). When using a human reference genome, nf-core/eager also now can give estimates of the biological sex of a given individual with Sex.DetEERRmine (@doi:10.1038/s41467-018-07483-5). A dedicated 'endogenous DNA' calculator (endorS.py) is also included to provide an approximate level of on-target DNA yield within a sample.

A major upgrade in contrast to the previous EAGER version is that the new pipeline supports processing of complex sequencing strategies for many samples. Given the large amount of sequencing often required to yield sufficient genome coverage from ancient DNA data, palaeogeneticists tend to use multiple (differently treated) libraries or sequencing runs. As an alternative to direct paths to FASTQ or BAM files, the pipeline can also accept a TSV file which includes file paths and additional metadata such as sample name, library name, sequencing lane, colour chemistry, and UDG treatment. This allows simultaneous processing and appropriate merging of heterogeneous data from multiple sequencing runs and/or libraries types.

Finally, the original EAGER tabular report format has been replaced with a much more extensive MultiQC (@doi:10.1093/bioinformatics/btw354) report. The original EAGER pipeline required users to look through many independent output directories and files to make full assessment of their sequencing data. Aggregation of all log files into a single interactive report will assist users in making fuller assessment of their sequencing and analysis runs. Most tools within nf-core/eager have a corresponding MultiQC module to ensure as complete an evaluation as possible.

## Accessibility

Alongside the portable new pipeline report, we have written extensive documentation on all parts of running and interpreting the output of the pipeline. Given that a large fraction of aDNA researchers come from fields outside computational biology, and thus may have limited computational training, we have written documentation that also gives guidance on how to interpret each section of the report, specifically in the context of NGS and aDNA. This includes schematic images of best practices or expected output that published under CC-BY licenses to allow for use in other training material. We hope this open-access resource will make the aDNA discipline more accessible to researchers new to the field, by providing practical and 'applied' knowledge as to how aDNA characteristics translate to downstream analyses.

The development of nf-core/eager in Nextflow and the nf-core initiative will also improve open-source community contributions to the pipeline. While Nextflow is written primarily in Groovy, the Nextflow DSL simplifies a number of concepts to an intermediate level that bioinformaticians without Java/Groovy experience can easily access (regardless of own programming language experience). Furthermore, Nextflow places ubiquitous and more widely known command-line interfaces, such as bash, in a prominent position within the code, rather than custom java code and classes. We hope this will motivate further bug fixes and feature contributions from the community to keep the pipeline updated with standard practises during a longer life-cycle. This will also be supported by the active and welcoming nf-core community who provide general guidance and advice on developing Nextflow and nf-core pipelines.
